File: Machine Learning A-Z/Part 4 - Clustering/Section 24 - K-Means Clustering/kmeans.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, [3,4]].values

from sklearn.cluster import KMeans
wcss=[]
for i in range(1,13):
    kmeans=KMeans(n_clusters=i,init='k-means++',n_init=10,max_iter=300,random_state=0)
    kmeans.fit(X)
    wcss.append(kmeans.inertia_)
    
    
plt.plot(range(1,13),wcss)
plt.title('Elbow method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.show()

kmeans=KMeans(n_clusters=5,init='k-means++',n_init=10,max_iter=300,random_state=0)
y_kmeans=kmeans.fit_predict(X)

colors=['red','blue','green','cyan','magenta']
titles=['Careful','Standard','Target','Careless','Sensible']
for i in range(0,5):
    plt.scatter(X[y_kmeans == i,0],X[y_kmeans == i, 1],s=100,c=colors[i],label=titles[i])

plt.scatter(kmeans.cluster_centers_[:,0],kmeans.cluster_centers_[:, 1],s=300,c='yellow',label='centroids')
plt.title('Kmeans')
plt.xlabel('Income')
plt.ylabel('Score')
plt.legend()
plt.show()

File: Machine Learning A-Z/Part 4 - Clustering/Section 24 - K-Means Clustering/kmeans.R
dataset = read.csv('Mall_Customers.csv')
x <- dataset[4:5]

set.seed(6)
wcss <- vector()

for(i in 1:10)
  wcss[i]<-sum(kmeans(x=x,centers = i)$withinss)

plot(1:10,wcss,type='b',main=paste('client clusters'),xlab='Number of clusters',ylab='WCSS')

set.seed(29)
model <- kmeans(x=x,centers = 5)
#values <- fitted(model,method=c("centers","classes"))

library(cluster)
clusplot(x,model$cluster,lines=0,shade=TRUE,color=TRUE,plotchar =FALSE,span = TRUE,main=paste('client clusters'),xlab='Annual Income',ylab='Spending Score')
